File: python/src/feb/week1/boj/수영장만들기_박재환.py
def solution():
    import sys
    input = sys.stdin.readline

    n,m = map(int, input().split())
    board = [list(map(int, input().strip())) for _ in range(n)]

    dx = [1,0,-1,0]
    dy = [0,1,0,-1]
    def check_water_h(h):
        from collections import deque
        q = deque()
        visited = [[False] * m for _ in range(n)]

        # 경계부터 탐색하며, 물이 찰 수 없는 위치를 시작 포인트로 기록
        for x in range(n):
            for y in range(m):
                if x == 0 or y == 0 or x == n-1 or y == m-1:
                    if board[x][y] < h:
                        visited[x][y] = True
                        q.append((x, y))

        while q:
            x, y = q.popleft()

            for dir in range(4):
                nx = x + dx[dir]
                ny = y + dy[dir]

                if nx < 0 or ny < 0 or nx >= n or ny >=m:
                    continue
                if visited[nx][ny] or board[nx][ny] >= h:
                    continue

                visited[nx][ny] = True
                q.append((nx, ny))

        water = 0
        for x in range(n):
            for y in range(m):
                if not visited[x][y] and board[x][y] < h:
                    water+=1
        return water

    result = 0
    for h in range(2,10):
        result += check_water_h(h)

    return result

if __name__ == '__main__':
    print(solution())

File: java/src/feb/week1/boj/수영장만들기_박재환.java
package feb.week1.boj;

import java.util.*;
import java.io.*;

public class 수영장만들기_박재환 {
    /**
     * 바깥으로 물이 세지 않는 칸들에, 주변 벽보다 낮은 만큼 물이 고인다.
     *
     * 갇혀있는 영역을 찾는다. => BFS / DFS
     */
    static BufferedReader br;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        init();
        br.close();
    }
    static StringTokenizer st;
    static int n, m;
    static int[][] map;
    static void init() throws IOException {
        st = new StringTokenizer(br.readLine().trim());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        map = new int[n][m];
        for(int x=0; x<n; x++) {
            String input = br.readLine().trim();
            for(int y=0; y<m; y++) map[x][y] = input.charAt(y) - '0';
        }

        int result = solution();
        System.out.println(result);
    }
    static int solution() {
        /**
         * 칸의 높이는 1 ~ 9
         * => 물의 높이 1 ~ 8
         */
        int result = 0;
        for(int height=2; height<10; height++) {
            result += checkWaterHeight(height);
        }
        return result;
    }
    static int[] dx = {0,1,0,-1};
    static int[] dy = {1,0,-1,0};
    static int checkWaterHeight(int height) {
        /**
         * 물이 채워질 수 없는 칸을 체크한다.
         */
        Queue<int[]> q = new ArrayDeque<>();
        boolean[][] visited = new boolean[n][m];

        // 시작 후보지점 정하기
        // 경계이거나, 물 높이보다 낮은 칸
        // => 물이 흘러나갈 수 있는 칸의 후보
        for(int x=0; x<n; x++) {
            for(int y=0; y<m; y++) {
                if(x == 0 || y == 0 || x == n-1 || y == m-1) {      // 경계인경우
                    if(!visited[x][y] && map[x][y] < height) {
                        visited[x][y] = true;
                        q.offer(new int[] {x, y});
                    }
                }
            }
        }
        // 확장
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];

            for(int dir=0; dir<4; dir++) {
                int nx = x + dx[dir];
                int ny = y + dy[dir];

                if(nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
                if(visited[nx][ny] || map[nx][ny] >= height) continue;

                visited[nx][ny] = true;
                q.offer(new int[] {nx, ny});
            }
        }

        int water = 0;
        for (int x = 0; x < n; x++) {
            for (int y = 0; y < m; y++) {
                if (map[x][y] < height && !visited[x][y]) {
                    water++;
                }
            }
        }
        return water;
    }
}
